sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/FileSystemAsyncClient.java

```diff
@@ -0,0 +1,611 @@
+     */
+    FileSystemAsyncClient(HttpPipeline pipeline, String url, DataLakeServiceVersion serviceVersion,
+        String accountName, String fileSystemName, BlobContainerAsyncClient blobContainerAsyncClient) {
+        this.azureDataLakeStorage = new DataLakeStorageClientBuilder()
+            .pipeline(pipeline)
+            .url(url)
+            .version(serviceVersion.getVersion())
+            .build();
+        this.serviceVersion = serviceVersion;
+
+        this.accountName = accountName;
+        this.fileSystemName = fileSystemName;
+        this.blobContainerAsyncClient = blobContainerAsyncClient;
+    }
+
+    /**
+     * Creates a new DataLakeFileAsyncClient object by concatenating fileName to the end of FileSystemAsyncClient's URL.
+     * The new DataLakeFileAsyncClient uses the same request policy pipeline as the FileSystemAsyncClient.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.file.datalake.FileSystemAsyncClient.getFileAsyncClient#String}
+     *
+     * @param fileName A {@code String} representing the name of the file.
+     * @return A new {@link DataLakeFileAsyncClient} object which references the file with the specified name in this
+     * file system.
+     */
+    public DataLakeFileAsyncClient getFileAsyncClient(String fileName) {
+        if (ImplUtils.isNullOrEmpty(fileName)) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("'fileName' can not be set to null"));
+        }
+        BlockBlobAsyncClient blockBlobAsyncClient = blobContainerAsyncClient.getBlobAsyncClient(fileName,
+            null).getBlockBlobAsyncClient();
+
+        return new DataLakeFileAsyncClient(getHttpPipeline(),
+            StorageImplUtils.appendToUrlPath(getFileSystemUrl(), Utility.urlEncode(Utility.urlDecode(fileName)))
+                .toString(), getServiceVersion(), getAccountName(), getFileSystemName(), fileName,
+            blockBlobAsyncClient);
+    }
+
+    /**
+     * Creates a new DataLakeDirectoryAsyncClient object by concatenating directoryName to the end of
+     * FileSystemAsyncClient's URL. The new DataLakeDirectoryAsyncClient uses the same request policy pipeline as the
+     * FileSystemAsyncClient.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.file.datalake.FileSystemAsyncClient.getDirectoryAsyncClient#String}
+     *
+     * @param directoryName A {@code String} representing the name of the directory.
+     * @return A new {@link DataLakeDirectoryAsyncClient} object which references the directory with the specified name
+     * in this file system.
+     */
+    public DataLakeDirectoryAsyncClient getDirectoryAsyncClient(String directoryName) {
+        if (ImplUtils.isNullOrEmpty(directoryName)) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("'directoryName' can not be set to null"));
+        }
```

Comment:
I think Objects.requireNonNull can make this more concise?

Reply:
Blobs does it this way, but issue created here https://github.com/Azure/azure-sdk-for-java/issues/6106